Openimageio to support more image formats?

Edi61 opened the pull request:
Hello,

I was missing the possibility to compare images which are in heic or avif format.
Therefore I have included Openimageio to support more image formats, since
FreeImage is not really developed anymore. Just a few lines of code was enough.
The implementation is surely not the best and the most efficent way, but it works for heic files.
To have an easier implementation I've used vcpkg to integrate Openimageio.

I would appreciate an improvement regarding this point.
What do you think regarding this point?

Regards,

sdottaka commented:
Thank you for your PR. Am I correct in understanding that you did not completely replace FreeImage with Openimageio because FreeImage supports image formats that OpenImageio does not support?

Edi61 commented:
No, it was more the implementation effort, since it would be a bigger change (Freeimage structure seems also be used inside winimerge) therefore I have not even tried to do.
FreeImage indeed supports some image formats which Openimageio is not supporting (Dr. Halo CUT, KOALA, Kodak PCD, Sun RAS,..) but they are very old formats which should not be relevant today and are even currently not supported by winimerge.

sdottaka commented:
Thank you for answering.

It looks like a lot more work needs to be done from the current PR as I have the following issues at the moment.

- When I tried this PR, running bootstrap-vcpkg.bat + vcpkg integrate install + buildbin.vs2019.cmd or opening WinIMerge.vs2022.sln in Visual Studio to build, I got the error " OpenImageIO/imageio.h" : No such file or directory. I was able to build it by manually specifying the vcpkg include directory and the library directory, but we do not want to add it manually. Also, we need to be able to build in a CI environment such as AppVeyor.

- There are many DLLs that depend on the following, totaling about 20 MB. Also, we need to modify BuildArc.cmd to get this DLL from the vcpkg directory for distribution. Also, the installer needs to be modified so that it can be used with WinMerge.

Iex-3_1.dll
libx265.dll
OpenImageIO.dll
OpenEXR-3_1.dll
heif.dll
jpeg62.dll
OpenImageIO_Util.dll
liblzma.dll
OpenEXRCore-3_1.dll
tiff.dll
de265.dll
Imath-3_1.dll
boost*.dll


FreeImage also seems to be able to handle HEIF and AVIF formats using the following, have you tried it?
  
https://github.com/mnaydenov/FreeImage-Sidecar

Edi61 commented:
Hi sdottaka,

since it's only a Draft PR, it's surely not complete. 
I did not jet consider overall intergration and
also other solutions then 2022.

Regarding Dlls:
OpenImageio can be also statically linked like Freeimage.
I have changed implementation (see last commit), so now
there is only one dll WinIMergeLib.dll as before.
Since both are using same libraries I was forced to remove
some of the plugins from FreeImage (Jpeg, Png, Tiff, OpenExr).
But it's not a real issue, now they are handled by OpenImageio
(with latest versions of there libraries).

Regarding Vcpkg:
There is no need to add anything regarding include files.
This is normally handled automatically by vcpkg.
Maybe something went wrong with the integrate install (was Visual Studio open?)?
Did you use Powershell to run execution? Did you have admin rights?
Did you you restart after integrate?
Maybe my former configuration of vcproj was wrong, I have updated it.

vcpkg is a widely used package manager supported by Microsoft
(https://github.com/microsoft/vcpkg) with a living community.
vcpkg itself is developed by continuous integration.
Also for projects using vcpkg continuous integration such as AppVeyor
is not a problem, it's already used by many projects.

Regarding FreeImage-Sidecar:
No, i haven't checked it, bit I would not like to investigate 
to a dead end (Old version, no activites, no future).

sdottaka commented:
Thank you for updating the PR.

My vcpkg problem went away, probably because I ran vcpkg integrate install in cmd.exe with admin rights.

However, now I'm getting the following link errors:

~~~
"C:\dev\winmerge\Externals\winimerge\WinIMerge.vs2022.sln" (rebuild target) (1) ->
"C:\dev\winmerge\Externals\winimerge\src\WinIMerge\winimerge.vcxproj.metaproj" (Rebuild target) (2) ->
"C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj.metaproj" (Rebuild target) (16) ->
"C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj" (Rebuild target) (17) ->
(Link target) ->
  tiff.lib(tif_win32.c.obj) : error LNK2005: _TIFFwarningHandler already defined in FreeImageLib.lib(PluginTIFF.obj) [C
:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  tiff.lib(tif_win32.c.obj) : error LNK2005: _TIFFerrorHandler already defined in FreeImageLib.lib(PluginTIFF.obj) [C:\
dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  tiff.lib(tif_win32.c.obj) : error LNK2005: _TIFFmalloc already defined in FreeImageLib.lib(PluginTIFF.obj) [C:\dev\wi
nmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  tiff.lib(tif_win32.c.obj) : error LNK2005: _TIFFcalloc already defined in FreeImageLib.lib(PluginTIFF.obj) [C:\dev\wi
nmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  tiff.lib(tif_win32.c.obj) : error LNK2005: _TIFFrealloc already defined in FreeImageLib.lib(PluginTIFF.obj) [C:\dev\w
inmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  tiff.lib(tif_win32.c.obj) : error LNK2005: _TIFFmemset already defined in FreeImageLib.lib(PluginTIFF.obj) [C:\dev\wi
nmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  tiff.lib(tif_win32.c.obj) : error LNK2005: _TIFFmemcpy already defined in FreeImageLib.lib(PluginTIFF.obj) [C:\dev\wi
nmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  tiff.lib(tif_win32.c.obj) : error LNK2005: _TIFFmemcmp already defined in FreeImageLib.lib(PluginTIFF.obj) [C:\dev\wi
nmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  tiff.lib(tif_win32.c.obj) : error LNK2005: _TIFFfree already defined in FreeImageLib.lib(PluginTIFF.obj) [C:\dev\winm
erge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  tiff.lib(tif_win32.c.obj) : error LNK2005: TIFFOpen already defined in FreeImageLib.lib(PluginTIFF.obj) [C:\dev\winme
rge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginTIFF.obj) : error LNK2001: unresolved external symbol "private: static union half::uif const *
 const half::_toFloat" (?_toFloat@half@@0QBTuif@1@B) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib
.vcxproj]
  FreeImageLib.lib(PluginJ2K.obj) : error LNK2001: unresolved external symbol opj_create_compress [C:\dev\winmerge\Exte
rnals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginJ2K.obj) : error LNK2001: unresolved external symbol opj_set_warning_handler [C:\dev\winmerge\
Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginJ2K.obj) : error LNK2001: unresolved external symbol opj_start_compress [C:\dev\winmerge\Exter
nals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginJ2K.obj) : error LNK2001: unresolved external symbol opj_destroy_codec [C:\dev\winmerge\Extern
als\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginJ2K.obj) : error LNK2001: unresolved external symbol opj_set_info_handler [C:\dev\winmerge\Ext
ernals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginJ2K.obj) : error LNK2001: unresolved external symbol opj_end_decompress [C:\dev\winmerge\Exter
nals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginJ2K.obj) : error LNK2001: unresolved external symbol opj_read_header [C:\dev\winmerge\External
s\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginJ2K.obj) : error LNK2001: unresolved external symbol opj_image_destroy [C:\dev\winmerge\Extern
als\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginJ2K.obj) : error LNK2001: unresolved external symbol opj_set_default_encoder_parameters [C:\de
v\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginJ2K.obj) : error LNK2001: unresolved external symbol opj_encode [C:\dev\winmerge\Externals\win
imerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginJ2K.obj) : error LNK2001: unresolved external symbol opj_setup_decoder [C:\dev\winmerge\Extern
als\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginJ2K.obj) : error LNK2001: unresolved external symbol opj_decode [C:\dev\winmerge\Externals\win
imerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginJ2K.obj) : error LNK2001: unresolved external symbol opj_setup_encoder [C:\dev\winmerge\Extern
als\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginJ2K.obj) : error LNK2001: unresolved external symbol opj_end_compress [C:\dev\winmerge\Externa
ls\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginJ2K.obj) : error LNK2001: unresolved external symbol opj_set_error_handler [C:\dev\winmerge\Ex
ternals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginJ2K.obj) : error LNK2001: unresolved external symbol opj_create_decompress [C:\dev\winmerge\Ex
ternals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginJ2K.obj) : error LNK2001: unresolved external symbol opj_set_default_decoder_parameters [C:\de
v\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "private: static short __cdecl half::conv
ert(int)" (?convert@half@@CAFH@Z) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: void __cdecl Imf_2_2::OutputFile
::writePixels(int)" (?writePixels@OutputFile@Imf_2_2@@QEAAXH@Z) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\W
inIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: void __cdecl Imf_2_2::OutputFile
::setFrameBuffer(class Imf_2_2::FrameBuffer const &)" (?setFrameBuffer@OutputFile@Imf_2_2@@QEAAXAEBVFrameBuffer@2@@Z) [
C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: virtual __cdecl Imf_2_2::OutputF
ile::~OutputFile(void)" (??1OutputFile@Imf_2_2@@UEAA@XZ) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMerg
eLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: __cdecl Imf_2_2::OutputFile::Out
putFile(class Imf_2_2::OStream &,class Imf_2_2::Header const &,int)" (??0OutputFile@Imf_2_2@@QEAA@AEAVOStream@1@AEBVHea
der@1@H@Z) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: class Imf_2_2::ChannelList::Cons
tIterator __cdecl Imf_2_2::ChannelList::end(void)const " (?end@ChannelList@Imf_2_2@@QEBA?AVConstIterator@12@XZ) [C:\dev
\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: class Imf_2_2::ChannelList::Cons
tIterator __cdecl Imf_2_2::ChannelList::begin(void)const " (?begin@ChannelList@Imf_2_2@@QEBA?AVConstIterator@12@XZ) [C:
\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: struct Imf_2_2::Channel const *
__cdecl Imf_2_2::ChannelList::findChannel(char const * const)const " (?findChannel@ChannelList@Imf_2_2@@QEBAPEBUChannel
@2@QEBD@Z) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: void __cdecl Imf_2_2::ChannelLis
t::insert(char const * const,struct Imf_2_2::Channel const &)" (?insert@ChannelList@Imf_2_2@@QEAAXQEBDAEBUChannel@2@@Z)
 [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: bool __cdecl Imf_2_2::Header::ha
sPreviewImage(void)const " (?hasPreviewImage@Header@Imf_2_2@@QEBA_NXZ) [C:\dev\winmerge\Externals\winimerge\src\WinIMer
geLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: class Imf_2_2::PreviewImage cons
t & __cdecl Imf_2_2::Header::previewImage(void)const " (?previewImage@Header@Imf_2_2@@QEBAAEBVPreviewImage@2@XZ) [C:\de
v\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: void __cdecl Imf_2_2::Header::se
tPreviewImage(class Imf_2_2::PreviewImage const &)" (?setPreviewImage@Header@Imf_2_2@@QEAAXAEBVPreviewImage@2@@Z) [C:\d
ev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: class Imf_2_2::ChannelList & __c
decl Imf_2_2::Header::channels(void)" (?channels@Header@Imf_2_2@@QEAAAEAVChannelList@2@XZ) [C:\dev\winmerge\Externals\w
inimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: class Imf_2_2::ChannelList const
 & __cdecl Imf_2_2::Header::channels(void)const " (?channels@Header@Imf_2_2@@QEBAAEBVChannelList@2@XZ) [C:\dev\winmerge
\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: class Imath_2_2::Box<class Imath
_2_2::Vec2<int> > const & __cdecl Imf_2_2::Header::dataWindow(void)const " (?dataWindow@Header@Imf_2_2@@QEBAAEBV?$Box@V
?$Vec2@H@Imath_2_2@@@Imath_2_2@@XZ) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: __cdecl Imf_2_2::Header::~Header
(void)" (??1Header@Imf_2_2@@QEAA@XZ) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: __cdecl Imf_2_2::Header::Header(
class Imath_2_2::Box<class Imath_2_2::Vec2<int> > const &,class Imath_2_2::Box<class Imath_2_2::Vec2<int> > const &,flo
at,class Imath_2_2::Vec2<float> const &,float,enum Imf_2_2::LineOrder,enum Imf_2_2::Compression)" (??0Header@Imf_2_2@@Q
EAA@AEBV?$Box@V?$Vec2@H@Imath_2_2@@@Imath_2_2@@0MAEBV?$Vec2@M@3@MW4LineOrder@1@W4Compression@1@@Z) [C:\dev\winmerge\Ext
ernals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "void __cdecl Imf_2_2::staticInitialize(v
oid)" (?staticInitialize@Imf_2_2@@YAXXZ) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: void __cdecl Imf_2_2::FrameBuffe
r::insert(char const * const,struct Imf_2_2::Slice const &)" (?insert@FrameBuffer@Imf_2_2@@QEAAXQEBDAEBUSlice@2@@Z) [C:
\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "int __cdecl Imf_2_2::globalThreadCount(v
oid)" (?globalThreadCount@Imf_2_2@@YAHXZ) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "protected: __cdecl Imf_2_2::IStream::ISt
ream(char const * const)" (??0IStream@Imf_2_2@@IEAA@QEBD@Z) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIM
ergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: virtual char * __cdecl Imf_2_2::
IStream::readMemoryMapped(int)" (?readMemoryMapped@IStream@Imf_2_2@@UEAAPEADH@Z) [C:\dev\winmerge\Externals\winimerge\s
rc\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: virtual bool __cdecl Imf_2_2::IS
tream::isMemoryMapped(void)const " (?isMemoryMapped@IStream@Imf_2_2@@UEBA_NXZ) [C:\dev\winmerge\Externals\winimerge\src
\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: virtual __cdecl Imf_2_2::IStream
::~IStream(void)" (??1IStream@Imf_2_2@@UEAA@XZ) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxp
roj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: void __cdecl Imf_2_2::InputFile:
:readPixels(int,int)" (?readPixels@InputFile@Imf_2_2@@QEAAXHH@Z) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\
WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: void __cdecl Imf_2_2::InputFile:
:setFrameBuffer(class Imf_2_2::FrameBuffer const &)" (?setFrameBuffer@InputFile@Imf_2_2@@QEAAXAEBVFrameBuffer@2@@Z) [C:
\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: class Imf_2_2::Header const & __
cdecl Imf_2_2::InputFile::header(void)const " (?header@InputFile@Imf_2_2@@QEBAAEBVHeader@2@XZ) [C:\dev\winmerge\Externa
ls\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: virtual __cdecl Imf_2_2::InputFi
le::~InputFile(void)" (??1InputFile@Imf_2_2@@UEAA@XZ) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLi
b.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: __cdecl Imf_2_2::InputFile::Inpu
tFile(class Imf_2_2::IStream &,int)" (??0InputFile@Imf_2_2@@QEAA@AEAVIStream@1@H@Z) [C:\dev\winmerge\Externals\winimerg
e\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "protected: __cdecl Imf_2_2::OStream::OSt
ream(char const * const)" (??0OStream@Imf_2_2@@IEAA@QEBD@Z) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIM
ergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: virtual __cdecl Imf_2_2::OStream
::~OStream(void)" (??1OStream@Imf_2_2@@UEAA@XZ) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxp
roj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: void __cdecl Imf_2_2::RgbaInputF
ile::readPixels(int,int)" (?readPixels@RgbaInputFile@Imf_2_2@@QEAAXHH@Z) [C:\dev\winmerge\Externals\winimerge\src\WinIM
ergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: void __cdecl Imf_2_2::RgbaInputF
ile::setFrameBuffer(struct Imf_2_2::Rgba *,unsigned __int64,unsigned __int64)" (?setFrameBuffer@RgbaInputFile@Imf_2_2@@
QEAAXPEAURgba@2@_K1@Z) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: virtual __cdecl Imf_2_2::RgbaInp
utFile::~RgbaInputFile(void)" (??1RgbaInputFile@Imf_2_2@@UEAA@XZ) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib
\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: __cdecl Imf_2_2::RgbaInputFile::
RgbaInputFile(class Imf_2_2::IStream &,int)" (??0RgbaInputFile@Imf_2_2@@QEAA@AEAVIStream@1@H@Z) [C:\dev\winmerge\Extern
als\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: __cdecl Imf_2_2::Channel::Channe
l(enum Imf_2_2::PixelType,int,int,bool)" (??0Channel@Imf_2_2@@QEAA@W4PixelType@1@HH_N@Z) [C:\dev\winmerge\Externals\win
imerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: __cdecl Imf_2_2::Slice::Slice(en
um Imf_2_2::PixelType,char *,unsigned __int64,unsigned __int64,int,int,double,bool,bool)" (??0Slice@Imf_2_2@@QEAA@W4Pix
elType@1@PEAD_K2HHN_N3@Z) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: void __cdecl Imf_2_2::RgbaOutput
File::writePixels(int)" (?writePixels@RgbaOutputFile@Imf_2_2@@QEAAXH@Z) [C:\dev\winmerge\Externals\winimerge\src\WinIMe
rgeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: void __cdecl Imf_2_2::RgbaOutput
File::setFrameBuffer(struct Imf_2_2::Rgba const *,unsigned __int64,unsigned __int64)" (?setFrameBuffer@RgbaOutputFile@I
mf_2_2@@QEAAXPEBURgba@2@_K1@Z) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: virtual __cdecl Imf_2_2::RgbaOut
putFile::~RgbaOutputFile(void)" (??1RgbaOutputFile@Imf_2_2@@UEAA@XZ) [C:\dev\winmerge\Externals\winimerge\src\WinIMerge
Lib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: __cdecl Imf_2_2::RgbaOutputFile:
:RgbaOutputFile(class Imf_2_2::OStream &,class Imf_2_2::Header const &,enum Imf_2_2::RgbaChannels,int)" (??0RgbaOutputF
ile@Imf_2_2@@QEAA@AEAVOStream@1@AEBVHeader@1@W4RgbaChannels@1@H@Z) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLi
b\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: __cdecl Imf_2_2::PreviewImage::~
PreviewImage(void)" (??1PreviewImage@Imf_2_2@@QEAA@XZ) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeL
ib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: __cdecl Imf_2_2::PreviewImage::P
reviewImage(unsigned int,unsigned int,struct Imf_2_2::PreviewRgba const * const)" (??0PreviewImage@Imf_2_2@@QEAA@IIQEBU
PreviewRgba@1@@Z) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "void __cdecl Iex_2_2::throwErrnoExc(void
)" (?throwErrnoExc@Iex_2_2@@YAXXZ) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: virtual char const * __cdecl Iex
_2_2::BaseExc::what(void)const " (?what@BaseExc@Iex_2_2@@UEBAPEBDXZ) [C:\dev\winmerge\Externals\winimerge\src\WinIMerge
Lib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: virtual __cdecl Iex_2_2::BaseExc
::~BaseExc(void)" (??1BaseExc@Iex_2_2@@UEAA@XZ) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxp
roj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: __cdecl Iex_2_2::BaseExc::BaseEx
c(class std::basic_stringstream<char,struct std::char_traits<char>,class std::allocator<char> > &)" (??0BaseExc@Iex_2_2
@@QEAA@AEAV?$basic_stringstream@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z) [C:\dev\winmerge\Externals\winimerge\
src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "public: __cdecl Iex_2_2::BaseExc::BaseEx
c(class Iex_2_2::BaseExc const &)" (??0BaseExc@Iex_2_2@@QEAA@AEBV01@@Z) [C:\dev\winmerge\Externals\winimerge\src\WinIMe
rgeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(PluginEXR.obj) : error LNK2001: unresolved external symbol "private: static unsigned short const * c
onst half::_eLut" (?_eLut@half@@0QBGB) [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(XTIFF.obj) : error LNK2001: unresolved external symbol _TIFFDataSize [C:\dev\winmerge\Externals\wini
merge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(J2KHelper.obj) : error LNK2001: unresolved external symbol opj_stream_set_read_function [C:\dev\winm
erge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(J2KHelper.obj) : error LNK2001: unresolved external symbol opj_stream_set_seek_function [C:\dev\winm
erge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(J2KHelper.obj) : error LNK2001: unresolved external symbol opj_stream_set_write_function [C:\dev\win
merge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(J2KHelper.obj) : error LNK2001: unresolved external symbol opj_stream_set_skip_function [C:\dev\winm
erge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(J2KHelper.obj) : error LNK2001: unresolved external symbol opj_stream_set_user_data_length [C:\dev\w
inmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(J2KHelper.obj) : error LNK2001: unresolved external symbol opj_stream_create [C:\dev\winmerge\Extern
als\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(J2KHelper.obj) : error LNK2001: unresolved external symbol opj_stream_set_user_data [C:\dev\winmerge
\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(J2KHelper.obj) : error LNK2001: unresolved external symbol opj_image_create [C:\dev\winmerge\Externa
ls\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  FreeImageLib.lib(J2KHelper.obj) : error LNK2001: unresolved external symbol opj_stream_destroy [C:\dev\winmerge\Exter
nals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]
  C:\dev\winmerge\Externals\winimerge\Build\x64\Release\WinIMerge\WinIMergeLib.dll : fatal error LNK1120: 77 unresolved
 externals [C:\dev\winmerge\Externals\winimerge\src\WinIMergeLib\WinIMergeLib.vcxproj]

    904 Warning(s)
    88 Error(s)

~~~

Edi61 commented:
Sorry, I have forgotten that freeimage have a seperate repo.
I have commited now also the removing of these plugins:
https://github.com/Edi61/freeimage/commit/b128cc55a9563d83fe4d62a1773918b5936d5a71
and update vcpkg to include all libs which are removed from freeimage.

sdottaka commented:
Thank you for the update.
Now I can build the x64 WinIMergeLib.dll. I like WinMerge to be compact, so the file size increase from 5.61MB to 11.6MB is really disappointing, but maybe I should accept it.

WinMerge still supports Windows XP and has ARM64 and ARM versions.
For this reason, I'm trying to build with VisualStudio 2017 or build the ARM64 or ARM version, but so far I haven't succeeded, so it seems necessary to investigate.

Edi61 commented:
I have reverted the changed for ARM/ARM64 since libheif seems not jet supporting ARM/ARM64 windows.
Now build should work again.
Ps.:
Just for information: Is ARM64/ARM really used? For which hardware ?

sdottaka commented:
Thank you for the PR update.

>    Just for information: Is ARM64/ARM really used? For which hardware?

I have been testing the functionality of the ARM64/ARM version of WinMerge on the Lenovo Yoga C630. Downloads for ARM64/ARM account for approximately 0.7% of the total.

Edi61 commented:
Forgotten to write: please update also freeimage
https://github.com/Edi61/freeimage/commit/fad67eef49661cbc45113d2d031faffa86987295

sdottaka commented:
I noticed that in Windows 11 (maybe even Windows 10) the Windows Imaging Component supports images with the heic extension. For this reason, I have added support for images with the heic extension using WIC in WinIMerge version 1.0.40.

At the moment, I cannot accept this PR due to some problems, such as not being able to compile it in Visual Studio 2017, but I will keep it in mind for future development.

Edi61 commented:
Hi sdottaka,

no problem, it's also fine for me. Would it be possible to add heic/heif extension in the open file window? Or maybe even to remove the extensional restriction completely and allow any file? Isn't the file format checked internally anyways?

sdottaka commented:
I have added file extensions containing heic, and heif in commit e56ae07a2ad351acd4db81c30eca4b92bfab72db.
